src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/BuildHostProcessManager.cs

```diff
@@ -0,0 +1,234 @@
+            UseShellExecute = false,
+            RedirectStandardInput = true,
+            RedirectStandardOutput = true,
+        };
+
+        AppendBuildHostCommandLineArgumentsAndRemoveEnvironmentVariables(processStartInfo);
+
+        var process = Process.Start(processStartInfo);
+        Contract.ThrowIfNull(process, "Process.Start failed to launch a process.");
+        return process;
+    }
+
+    private void AppendBuildHostCommandLineArgumentsAndRemoveEnvironmentVariables(ProcessStartInfo processStartInfo)
+    {
+        if (_binaryLogPath is not null)
+        {
+            processStartInfo.ArgumentList.Add("--binlog");
+            processStartInfo.ArgumentList.Add(_binaryLogPath);
+        }
+
+        // MSBUILD_EXE_PATH is read by MSBuild to find related tasks and targets. We don't want this to be inherited by our build process, or otherwise
+        // it might try to load targets that aren't appropriate for the build host.
+        processStartInfo.Environment.Remove("MSBUILD_EXE_PATH");
+    }
+
+    private static readonly XmlReaderSettings s_xmlSettings = new()
+    {
+        DtdProcessing = DtdProcessing.Prohibit,
+    };
+
+    private static BuildHostProcessKind GetKindForProject(string projectFilePath)
+    {
+        // This implements the algorithm as stated in https://github.com/dotnet/project-system/blob/9a761848e0f330a45e349685a266fea00ac3d9c5/docs/opening-with-new-project-system.md;
+        // we'll load the XML of the project directly, and inspect for certain elements.
+        XDocument document;
+
+        // Read the XML, prohibiting DTD processing due the the usual concerns there.
+        using (var fileStream = new FileStream(projectFilePath, FileMode.Open, FileAccess.Read))
+        using (var xmlReader = XmlReader.Create(fileStream, s_xmlSettings))
+            document = XDocument.Load(xmlReader);
+
+        // If we don't have a root, doesn't really matter which. This project is just malformed.
+        if (document.Root == null)
+            return BuildHostProcessKind.NetCore;
+
+        // Look for SDK attribute on the root
+        if (document.Root.Attribute("Sdk") != null)
+            return BuildHostProcessKind.NetCore;
+
+        // Look for <Import Sdk=... />
+        if (document.Root.Elements("Import").Attributes("Sdk").Any())
+            return BuildHostProcessKind.NetCore;
+
+        // Look for <Sdk ... />
+        if (document.Root.Elements("Sdk").Any())
+            return BuildHostProcessKind.NetCore;
+
+        // Looking for PropertyGroups that contain TargetFramework or TargetFrameworks nodes
+        var propertyGroups = document.Descendants("PropertyGroup");
+        if (propertyGroups.Elements("TargetFramework").Any() || propertyGroups.Elements("TargetFrameworks").Any())
+            return BuildHostProcessKind.NetCore;
+
+        // Nothing that indicates it's an SDK-style project, so use our .NET framework host
+        return BuildHostProcessKind.NetFramework;
```

Comment:
Reminder for me: until we have Mono support, this only return NetFramework on Windows, since it definitely won't work on Mac/Linux.